### ISSUE TITLE
Create a page for Specifying Open Liberty bootstrap properties

### ISSUE THREAD

**Issue by dmuelle** [OpenLiberty/docs#5708]
similar to https://www.ibm.com/docs/en/was-liberty/base?topic=liberty-specifying-bootstrap-properties, but review whether the task format is the best framework for this information and work with an SME to determine the relevant bits, as well as anything that might be missing.

SME- TBD

**Comment by dmuelle** [comment]
need to look at what is already in OL to avoid duplication, make sure a new page is warranted, and if so, reorganize existing info as needed.

**Comment by dmuelle** [comment]
from @NottyCode in #5707 

> I also cannot find a page that does talk about what properties in bootstrap.properties have special meanings. Doing a google I find [this page](https://www.ibm.com/docs/en/was-liberty/base?topic=liberty-specifying-bootstrap-properties) which does cover this kind of material, but I couldn't find equivalent in Open Liberty docs. It would be good to review that document to see what is/should be covered in Open Liberty docs.

**Comment by ramkumar-k-9286** [comment]
Hi David @dmuelle 

Draft link added: https://docs-draft-openlibertyio.mqj6zf7jocq.us-south.codeengine.appdomain.cloud/docs/latest/bootstrap-properties.html

Noticed an issue with the Nav. Will fix that.

Regards
Ramkumar

**Comment by dmuelle** [comment]
some initial feedback-

- run acrolinx
- try not to have paragraphs with only one sentence unless you are introducing an example and want to set the intro statement apart
- combine the two OSIGI properties under a single heading
- improve the heading names. I dont think it matters whether a property is predefined or not since the bootstrap properties file doesn't exist unless you create it. The process for the user is the same whether or not the property is predefined. You could mention in the description but make the headings focus on what the properties do. eg ~~Custom properties to set the command port~~ ---> Set the command port
- dont need two tables in the command port section, just mention that 0 is the default the first time
- table needs an intro sentence.
- dont need numbered bullets in the HTTP port section

> You must create this file in the server directory, which also contains the configuration root file server.xml. 

--->

>  You must create this file in the server configuration directory, which also contains the `server.xml` file. 

---

> If you update the bootstrap.properties file, you must restart the server for the changes to take effect.

--->

> If you update the `bootstrap.properties` file, you must restart the server for the changes to take effect. However, when Open Liberty is running in dev mode, it automatically detects changes to the `bootstrap.properties` file and restarts the server to apply them.

**Comment by ramkumar-k-9286** [comment]
Hi David @dmuelle

Suggested corrections made.

Draft link: https://docs-draft-openlibertyio.mqj6zf7jocq.us-south.codeengine.appdomain.cloud/docs/latest/reference/bootstrap-properties.html


Regards
Ramkumar

**Comment by ramkumar-k-9286** [comment]
Hi Alaine @acdemyers 

 Please have a look at this draft document
 Draft link: https://docs-draft-openlibertyio.mqj6zf7jocq.us-south.codeengine.appdomain.cloud/docs/latest/reference/bootstrap-properties.html
 
 Had some questions regarding the **default ports for web applications** section. -> How does the `httpEndpoint` specified in the 'server.xml` file relate to the bootstrap.properties file?
 
 Could you also review the entire doc and suggest changes or corrections? This doc is being ported from https://www.ibm.com/docs/en/was-liberty/base?topic=liberty-specifying-bootstrap-properties, .
 
 Please let us know if the new draft document is fine or needs changes 
 
 Regards,
Ramkumar

CC @dmuelle

**Comment by acdemyers** [comment]
Hi Ramkumar, @ramkumar-k-9286 

I don't handle the ports that is done by the channel.  I've asked Lori @loriadi if she or somebody on her team can take a look.

Alaine

**Comment by loriadi** [comment]
I'm looking at the host and and port variables.

- "Had some questions regarding the default ports for web applications section. -> How does the httpEndpoint specified in the 'server.xml` file relate to the bootstrap.properties file?"

--The http and https port properties defined in bootstrap.properties can be used as variables on the httpendpoint in server.xml.

- I'm not sure about  the word "adapter" used in this section.  I see in other places we use "network interface".
https://openliberty.io/docs/latest/reference/config/httpEndpoint.html

On number 1., we say these are the properties to set in the bootstrap.properties file.  On number 2. , can we say use "these" properties in the server.xml ...

I'm not familiar with the command port property, but I see there's no example provided.

**Comment by dmuelle** [comment]
Thanks @loriadi - to make sure I understand correctly..

I think some of the confusion is that "default" is being used in two different senses here. One the one hand, We have "default" http/s ports (9080,9444) which is what the server uses if you dont specify anything else for a given httpendpoint. On the other, you can specify `default.http/s.port` variables to change those values so that they become the "default" for any endpoint that specifies them- but really they're just like any custom variable you specify in the bootstrap properties. They only get used as a default if you specify the bootstrap-defined variable in the endpoint definition- is that correct?

**Comment by dmuelle** [comment]
for the command port- we will be adding an example like the other sections on the page

**Comment by dmuelle** [comment]
One other question- the command port section describes how to disable the command port, but then says not to do it. Is this something we for sure want to document? Are there cases where users will need to disable the command port or is that more of an internal thing?

**Comment by loriadi** [comment]
> Thanks @loriadi - to make sure I understand correctly..
> 
> I think some of the confusion is that "default" is being used in two different senses here. One the one hand, We have "default" http/s ports (9080,**9443**) which is what the server uses if you dont specify anything else for a given httpendpoint. On the other, you can specify `default.http/s.port` variables to change those values so that they become the "default" for any endpoint that specifies them- but really they're just like any custom variable you specify in the bootstrap properties. They only get used as a default if you specify the bootstrap-defined variable in the endpoint definition- is that correct?

Yes, that's correct and confusing.  Maybe we should be saying default and custom for the two cases.

If I create a basic bare bones server using "server create test", the server.xml that gets created has this for the httpendpoint:
<httpendpoint id="defaultHttpEndpoint" 
       httpPort="9080"
       httpsPort="9443" />
       
If I change the server.xml endpoint to the following and add those properties to the bootstrap.properties file then the custom properties are used to set the http/s port.:
<httpEndpoint id="defaultHttpEndpoint"
      httpPort="${default.http.port}"
      httpsPort="${default.https.port}" />
  
 Also, the host="*" discusssion may be out of scope for this topic.  Most customers would add host= "*" but it doesn't have anything to do with the bootstrap.properties and it isn't set by default when a server is created.  Because most customers would use it, maybe it's fine to leave it?  Either way it's fine, but it causes the text to explain it which may be less direct.

**Comment by ramkumar-k-9286** [comment]
Hi David @dmuelle 

Draft link with the changes suggested. Please have a look at them. 

Please have a look at this draft document
Draft link: https://docs-draft-openlibertyio.mqj6zf7jocq.us-south.codeengine.appdomain.cloud/docs/latest/reference/bootstrap-properties.html

Regards,
Ramkumar

**Comment by dmuelle** [comment]
Looks good- I think http section still needs a little work to be more precise and clear.

You can use the bootstrap.properties file to create custom default HTTP and HTTPS properties.
-->
You can define custom variables in the `bootstrap.properties` file that specify custom default values for HTTP and HTTPS ports. These values are applied to any config:httpEndpoint[display=HTTP endpoint configuration] that specifies the variables in the `server.xml` file.

---

Specify the default.http.port and default.https.port properties in the bootstrap.properties file. If you do not specify these properties, the default HTTP port is 9080 and the default HTTPS port is 9443.

--->

The default value for the HTTP port is 9080 and the default value for the HTTPS port is 9443. You can create custom defaults to replace these values by specifying the default.http.port and default.https.port properties in the bootstrap.properties file, as shown in the following example.

---

You can use the default.http.port and default.https.port properties specified in the bootstrap.properties file as custom properties from the server.xml configuration file.
--->
After you define these properties in the `bootstrap.properties` file, you can use them as variables to define the port values for your HTTP endpoint configuration in the `server.xml` file, as shown in the following example.

---

For more information about how variables are applied in the server configuration, see Variable substitution precedence. (and check the link, broken in the draft)

**Comment by dmuelle** [comment]
## Peer review

The short description seems repetitive:

> Bootstrap properties initialize the runtime environment for a particular server. Bootstrap properties are attributes that affect the configuration and initialization of the runtime core.

maybe

> Bootstrap properties initialize the runtime environment for a particular server. You can set these properties to customize  the configuration and initialization of the runtime core.

---

Italicize the variable in the path:

By default, the server directory is `usr/servers/server_name`.

--->

By default, the server configuration directory is `usr/servers/ _server_name_ `.

Doesnt show up in MArkdown but in asciidoc, that syntax will italicize _server_name_

---

However, when Open Liberty is running in dev mode...
--->
However, when Open Liberty is running in xref:ROOT:development-mode.adoc[dev mode]...

---

To enable binary logging
--->
To enable xref:ROOT:log-trace-configuration.adocl#binary[binary logging]

---

This code snippet needs proper indenting (this was a problem from the original doc...)

```
<httpEndpoint id="defaultHttpEndpoint"
      httpPort="${default.http.port}"
      httpsPort="${default.https.port}" />
```

also need to set [source,xml] for that snippet

---

this makes it sound like you have to actively set the port, which wont be true in the majority of cases


Set the command port to enable the server script to communicate with the running Open Liberty server and request certain operations...
--->
The command port enables the server script to communicate with the running Open Liberty server and request certain operations...

---

missing period and also imprecise-

The user-specified values for the command.port are 1 to 65535
--->
The  valid range of user-specified values for the command.port is 1 to 65535.

---

You are discouraged from disabling the command port. The value for disabling the command.port is -1.

--->

You can disable the command port by setting the `command.port` property to `-1`. However, this setting is not recommended.

---

something not quite right here

You can **increase** the server start wait time beyond the product default setting by specifying the server.start.wait.time property to the boostrap.properties file.

The following example sets the server start time to 25 seconds.

The description says you can increase, but the example decreases (25 vs 30). I think we should change the example to be 35.

**Comment by ramkumar-k-9286** [comment]
Hi David @dmuelle

Peer Review comments have been incorporated. Please have a look at them.

Please have a look at this draft document
Draft link: https://docs-draft-openlibertyio.mqj6zf7jocq.us-south.codeengine.appdomain.cloud/docs/latest/reference/bootstrap-properties.html

Regards,
Ramkumar

**Noticed and issue with the Binary Logging, variable-substitution link. Will fix that.**

**Comment by ramkumar-k-9286** [comment]
Hi Lori @loriadi 

All updates and review comments have been incorporated.

Could you go ahead and have a look at the document?

Draft link: https://docs-draft-openlibertyio.mqj6zf7jocq.us-south.codeengine.appdomain.cloud/docs/latest/reference/bootstrap-properties.html

If you are happy with the changes, please add the `Developer Reviewed` label to this issue.

if any changes are needed, please add a comment to the issue.

Regards,
Ramkumar

CC @dmuelle

**Comment by ramkumar-k-9286** [comment]
Hi David @dmuelle 

Staging Link: https://docs-staging-openlibertyio.mqj6zf7jocq.us-south.codeengine.appdomain.cloud/docs/latest/reference/bootstrap-properties.html

Regards,
Ramkumar

**Comment by dmuelle** [comment]
published with 22.0.0.12, closing as completed
